component/file_cache/file_cache.go

```diff
@@ -983,42 +996,59 @@ func (fc *FileCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Hand
 
 // CloseFile: Flush the file and invalidate it from the cache.
 func (fc *FileCache) CloseFile(options internal.CloseFileOptions) error {
-	log.Trace("FileCache::CloseFile : name=%s, handle=%d", options.Handle.Path, options.Handle.ID)
+	// Lock the file so that while close is in progress no one can open the file again
+	flock := fc.fileLocks.Get(options.Handle.Path)
+	flock.Lock()
+
+	if !fc.lazyWrite {
+		// Sync close is called so wait till the upload completes
+		return fc.closeFileInternal(options, flock)
+	}
+
+	// Async close is called so schedule the upload and return here
+	fc.asyncClose.Add(1)
+	go fc.closeFileInternal(options, flock) //nolint
+	return nil
```

Comment:
How to know if the CloseFile returned error or not in case of lazy write?

Reply:
For user application is will always pass and only logs will tell whether file upload failed or not.
We need to think on if anything fails can we re-push the upload request and try again.